Compile error "too many sections" on windows/msys2/mingw-w64/gcc

Compile on Windows 7, in the msys2 environment, with gcc compiler and mingw-w64 runtime headers/libs got me the error `too many sections` and a compile fail.

Here is some console io:

```
realz@realz-PC MINGW64 ~/include-what-you-use/build
$ make
Scanning dependencies of target include-what-you-use
[  7%] Building CXX object CMakeFiles/include-what-you-use.dir/iwyu.cc.obj
In file included from K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use/iwyu_ast_util.h:19:0,
                 from K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use/iwyu.cc:104:
K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use/port.h:68:0: warning: "NOMINMAX" redefined
 #define NOMINMAX  // Prevent Windows headers from redefining min/max.
 ^
In file included from K:/realz/dump/mingw-w64/msys64/mingw64/include/c++/5.2.0/x86_64-w64-mingw32/bits/c++config.h:482:0,
                 from K:/realz/dump/mingw-w64/msys64/mingw64/include/c++/5.2.0/utility:68,
                 from K:/realz/dump/mingw-w64/msys64/mingw64/include/c++/5.2.0/algorithm:60,
                 from K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use/iwyu.cc:93:
K:/realz/dump/mingw-w64/msys64/mingw64/include/c++/5.2.0/x86_64-w64-mingw32/bits/os_defines.h:45:0: note: this is the location of the previous definition
 #define NOMINMAX 1
 ^
K:/realz/dump/mingw-w64/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/include-what-you-use.dir/iwyu.cc.obj: too many sections (38804)
K:\realz\dump\mingw-w64\msys64\tmp\ccY0rz3g.s: Assembler messages:
K:\realz\dump\mingw-w64\msys64\tmp\ccY0rz3g.s: Fatal error: can't write CMakeFiles/include-what-you-use.dir/iwyu.cc.obj: File too big
K:/realz/dump/mingw-w64/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/5.2.0/../../../../x86_64-w64-mingw32/bin/as.exe: CMakeFiles/include-what-you-use.dir/iwyu.cc.obj: too many sections (38804)
K:\realz\dump\mingw-w64\msys64\tmp\ccY0rz3g.s: Fatal error: can't close CMakeFiles/include-what-you-use.dir/iwyu.cc.obj: File too big
CMakeFiles/include-what-you-use.dir/build.make:62: recipe for target 'CMakeFiles/include-what-you-use.dir/iwyu.cc.obj' failed
make[2]: *** [CMakeFiles/include-what-you-use.dir/iwyu.cc.obj] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/include-what-you-use.dir/all' failed
make[1]: *** [CMakeFiles/include-what-you-use.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

The solution (after googling the issue) was to pass in `Wa,-mbig-obj` to the CXX_FLAGS:

```
realz@realz-PC MINGW64 ~/include-what-you-use/build
$ cmake -G"MSYS Makefiles" .. -DIWYU_LLVM_INCLUDE_PATH=/home/realz/.local/include/ -DIWYU_LLVM_LIB_PATH=/home/realz/.local/lib -DCMAKE_CXX_FLAGS=-Wa,-mbig-obj  -DCMAKE_VERBOSE_MAKEFILE=1
```

Not sure if you want to put this conditionally in the CMakeLists.txt, but I'll just leave this here for any future users who come across the issue.

@kimgr Any chance a new error was introduced?

> ```
K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use-fork/include-what-you-use/iwyu_ast_util.cc: In function 'std::map<const clang::Type*, const clang::Type*> include_what_you_use::GetTplTypeResugarMapForFunctionExplicitTplArgs(const clang::FunctionDecl*, const clang::TemplateArgumentListInfo&)':
K:/realz/dump/mingw-w64/msys64/home/realz/include-what-you-use-fork/include-what-you-use/iwyu_ast_util.cc:716:59: error: 'const class clang::TemplateArgumentListInfo' has no member named 'arguments'
   for (const TemplateArgumentLoc& loc : explicit_tpl_list.arguments()) {
                                                           ^
CMakeFiles/include-what-you-use.dir/build.make:86: recipe for target 'CMakeFiles/include-what-you-use.dir/iwyu_ast_util.cc.obj' failed
make[2]: *** [CMakeFiles/include-what-you-use.dir/iwyu_ast_util.cc.obj] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/include-what-you-use.dir/all' failed
make[1]: *** [CMakeFiles/include-what-you-use.dir/all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -89,6 +89,11 @@ add_executable(include-what-you-use
   iwyu_verrs.cc
 )
 
+if( MINGW )
+  # Work around 'too many sections' error with MINGW/GCC
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wa,-mbig-obj")
+endif()
+
 if( MSVC )
   # Disable warnings for IWYU, and disable exceptions in MSVC's STL.
   add_definitions(

